docker/src/main/java/clocker/docker/location/strategy/basic/GroupPlacementStrategy.java

```diff
@@ -0,0 +1,182 @@
+
+import com.google.common.annotations.Beta;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.reflect.TypeToken;
+import com.google.common.util.concurrent.Monitor;
+import com.google.common.util.concurrent.Uninterruptibles;
+
+import org.apache.brooklyn.api.entity.Entity;
+import org.apache.brooklyn.config.ConfigKey;
+import org.apache.brooklyn.core.config.ConfigKeys;
+import org.apache.brooklyn.util.collections.MutableList;
+import org.apache.brooklyn.util.collections.MutableMap;
+import org.apache.brooklyn.util.core.flags.SetFromFlag;
+import org.apache.brooklyn.util.core.mutex.WithMutexes;
+import org.apache.brooklyn.util.exceptions.Exceptions;
+import org.apache.brooklyn.util.time.Duration;
+
+/**
+ * A {@link DockerAwarePlacementStrategy strategy} that requires entities with
+ * the same parent to use the same host.
+ *
+ * Can be configured to require exclusive use of the host with the
+ * {@link #REQUIRE_EXCLUSIVE exclusive} ({@code docker.constraint.exclusive})
+ * option set to {@code true}; normally {@code false}.
+ *
+ * @since 1.1.0
+ */
+@Beta
+@ThreadSafe
+public class GroupPlacementStrategy extends AbstractDockerPlacementStrategy implements WithMutexes {
+
+    private static final Logger LOG = LoggerFactory.getLogger(GroupPlacementStrategy.class);
+    private static final Object LOCK = new Object[0];
+
+    @SetFromFlag("exclusive")
+    public static final ConfigKey<Boolean> REQUIRE_EXCLUSIVE = ConfigKeys.newBooleanConfigKey(
+            "docker.constraint.exclusive",
+            "Whether the Docker host must be exclusive to this application; by default other applications can co-exist",
+            Boolean.FALSE);
+
+    @SetFromFlag("timeout")
+    public static final ConfigKey<Duration> STRATEGY_TIMEOUT = ConfigKeys.newDurationConfigKey(
+            "docker.strategy.timeout",
+            "How long to wait for other entities using the strategy",
+            Duration.minutes(20));
+
+    public static final ConfigKey<Map<String, Monitor>> MONITOR_MAP = ConfigKeys.newConfigKey(
+            new TypeToken<Map<String, Monitor>>() { },
+            "groupPlacementStrategy.map.monitors",
+            "A mapping from application IDs to monitors; used to synchronize threads during strategy execution.");
+
+    @Override
+    public List<DockerHostLocation> filterLocations(List<DockerHostLocation> locations, Entity entity) {
+        if (locations == null || locations.isEmpty()) {
+            return ImmutableList.of();
+        }
+        if (getDockerInfrastructure() == null) config().set(DOCKER_INFRASTRUCTURE, Iterables.getLast(locations).getDockerInfrastructure());
+        List<DockerHostLocation> available = MutableList.copyOf(locations);
+        boolean requireExclusive = config().get(REQUIRE_EXCLUSIVE);
+
+        try {
+            acquireMutex(entity.getApplicationId(), "Filtering locations for " + entity);
```

Comment:
I find it confusing the way that `releaseMutex()` is called in the `DockerLocation` code. Deserves a comment here to explain that. Not until you search for `releaseMutex` does one understand how this code can work.